Hot rod dialect v4

Sanne opened the pull request:
Some pending tasks are documented as sub-tasks on JIRA:
 - https://hibernate.atlassian.net/browse/OGM-353

(Although, I've started including some of the sub-tasks already since opening this PR; next working on the missing Sequencer tests and then documentation)

WildFly modules included
  - https://hibernate.atlassian.net/browse/OGM-1160


# Launching / Exploring tests

As usual, there's a `org.hibernate.ogm.datastore.infinispanremote.utils.InfinispanRemoteBackendTckHelper` to help you start a test from OGM/core on this new dialect.
It includes a `@ClassRule` which will start an Infinispan Server as needed, or you can launch it via the static `main()` method to keep the server running (manually).

The tests which are specific to this module will use the same JUnit Rule so the database is auto-started as needed.

When running it via Maven, the Surefire plugin is instructed to activate the testing listener `org.hibernate.ogm.datastore.infinispanremote.utils.HotrodServerLifecycle`, which will keep the Infinispan Server running for the duration of the tests, and rest its state as needed.

I like this new approach, as the test execution is very fast and the debugging / development sessions are more practical.. we might consider evolving other modules to use a similar approach.

For the Arquillian tests, the magic is handled by an Arquillian extension which invokes the same JUnit helpers as before, but in the right timing as it needs to happen before the deployment: `org.hibernate.ogm.test.integration.infinispanremote.HotRodServerLifecycleManager` 
 - https://github.com/Sanne/hibernate-ogm/blob/bed210d92f8a3cf3b4033b8971b5206bef0665d0/integrationtest/src/test/java/org/hibernate/ogm/test/integration/infinispanremote/HotRodServerLifecycleManager.java

# Schema & Encoders

This dialect needs a full schema to be deployed to the Server before any I/O can happen.

We already discussed the shape and strategy of the schema, but to remind you it's 1:1 of how ORM would map it to RDBMs Tables, I'll highlight the exceptions to this rule in the following sections.

The schema build process is driven by `org.hibernate.ogm.datastore.infinispanremote.impl.ProtobufSchemaInitializer`.

The core element representing a "Table" mapping is
`org.hibernate.ogm.datastore.infinispanremote.impl.schema.TableDefinition`
 - https://github.com/Sanne/hibernate-ogm/blob/bed210d92f8a3cf3b4033b8971b5206bef0665d0/infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/schema/TableDefinition.java

These "TableDefinitions" are generated upfront during boot time, and are immutable after that.
They use "ProtofieldWriterSet" which encapsulates an ordered set of field definitions, each with its:
 - ORM column name
 - Protobuf field name
 - Encoder/Decoder references

So a single _TableDefinition_ is able to both export itself in string format to generate the schema DDL fragment, but is also containing the function to decode / encode the payload to the underlying buffers.

The actual driver for encoding / decoding is `org.hibernate.ogm.datastore.infinispanremote.impl.protostream.ProtoDataMapper` , so the main dialect will interact with the _ProtoDataMapper_

## Capturing the generated Schema

There's a pluggable SPI `org.hibernate.ogm.datastore.infinispanremote.spi.schema.SchemaCapture` to intercept / capture the generated protobuf schema. This is controlled by a configuration option, and the default is to log the generated schema but in some tests I make use of it to validate the schema generation.
The user could also make use of this to override the schema generation and deploy a custom one (not that I'd suggest doing that as our encoders wouldn't be affected).

## Sequences and Table based generators

A sequence is mapped to a key / value, where the key is the sequence name, and stored in a Cache having the sequence name.

When using a table generator, ORM can store multiple "lines" called segments; in this case we consider the Table name as Cache name and each combination segment name part of the ID for the key/value store.

What's interesting is that, no matter if ORM ops for Sequences or Table strategy, we're actually storing in the same way.

The actual sequences are implemented via versioned CAS operations:
 - https://github.com/Sanne/hibernate-ogm/blob/bed210d92f8a3cf3b4033b8971b5206bef0665d0/infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/sequences/HotRodSequencer.java

## Bags & unordered collections

Where a collection is of a "bag" type, ORM would map this to a table lacking a primary key, I guess primarily to support duplicate entries.
In a Key/Value store, the key is hardly optional so I have found no way to map this, and noticed the Cassandra dialect doesn't support it either.
The good news is that we detect this early on (at bootstrap, during schema generation) and throw a helpful exception.

# Versioning and Optimistic locking

Optimistic locking is used as an implementation detail by the Sequence generator, but it's not used as a general storage feature by the dialect.

There is some commented-out code in the `InfinispanRemoteDialect` which I believe should make it easy to implement proper optimistic locking in the near future, but I suspect this requires the fix on number of operations by @gsmet first.

# InfinispanRemoteDialect

The main Dialect driver is org.hibernate.ogm.datastore.infinispanremote.InfinispanRemoteDialect<EK, AK, ISK>.

Among the optional features, it supports `MultigetGridDialect`.

The Dialect interacts with the usual Tuples from OGM but internally uses the pairs 
-  `ProtostreamPayload` to represent a ready-to-be-encoded Value
- `ProtostreamId` to represent a ready-to-be-encoded Key

These are the only types recognized by the two-way encoders so actual storage operations need to be handled using these payload objects.
Finally, all actual Cache IO operations have to be executed within the context of a "mapper.withinCacheEncodingContext( lambda )" so that the DataMapper can setup the right encoding scheme for the specific Cache. This is to workaround limitation in how the Protostream encoders are configured, as normally they have to be setup 1:1 to class types.. more on this on JIRA s to follow up so we can cleanup the ThreadLocal I'm using for this workaround.

gunnarmorling commented:
Hey @Sanne, awesome to see this PR; looking forward to have the remote support in finally!

I was wondering a bit how to approach this PR; do you think you could provide some pointers to the design from a more high-level point of view, important things/classes to look at and things like that? The change is huge so I'd like to make sure we can make the review efficient by getting a good understanding of the overall picture and focus on the most important things. Thanks!

Sanne commented:
> I was wondering a bit how to approach this PR

You're on holidays, isn't that a good hint on how to approach this? ;-)

But nice idea, I'll write a mini-guide in the PR description.

gunnarmorling commented:
> I'll write a mini-guide in the PR description

Cool, thanks!

emmanuelbernard commented:
I'll look at it a bit but do not count on me for a full review. So someone else, please review Sanne's work.

emmanuelbernard commented:
I think I would prefer to map associations within the entity itself and not as a separate entry by default. First off, it's not possible today due to a limitation of Infinispan's ProtoStream: @Sanne needs to keep pushing the Infinispan team to address the remaining issues (this one and the ability to push schema come to mind).

But assuming we can, we did discuss it a bit with Sanne, but let me try and summarize my points:

By having the association data (ids in this case) within the entity entry, you allow more queries (in a wide sense) to be executed by Infinispan. In particular all of the queries requiring both an entity property and an association state. Queries not involving both can be handled by a separate query to the association cache.
This is particularly acute for `@ElementCollection` which naturally fit within the entity. Embedding the state is how Teiid maps things into Infinispan (see Van).

Granted, you need the option to store the association state in a separate entry (like we do for Mongo) in case the entry becomes big. We should reuse the existing option annotations.

By Infinispan query, I mean all means of getting the data except look up: dist exec, distributed streams, spark, Infinispan query, Full-text search etc.

TODO: capture that discussion in a JIRA to execute on it once Infinispan fix the limitations.

emmanuelbernard commented:
== Bags & unordered collections

The way you describe the problem made me wonder how you store this info @Sanne.
Is there one association cache for all possible associations? Or one association cache per association type.
And that question aside, how do you store entries? one key per entry in the association or rather one key per association and the value contains a collection of entries?
If the former, how do you retrieve all elements of an association (say to satisfy a a.`getBs()` call)?
I imagine in this case that you could give synthetic UUID keys as neither order matter and Hibernate ORM does ask to remove all entries matching the columns (i.e. all the duplicates) and recreates the expected amount of duplicates to match the new state. For example if you have 3 times `[nick:joe, age:18]` and remove one, Hibernate ORM asks to remove all entries containing `[nick:joe, age:18]` and then asks you to add two entries `[nick:joe, age:18]`. Additions of course are not a problem. So you could walk over all UUID keys and remove the duplicates by value.

If the latter (value is a collection of entries), you can apply the same logic.

gunnarmorling commented:
> Sanne added some commits on *18 Dec 2015*

Holy moly ... :)

gunnarmorling commented:
> I think I would prefer to map associations within the entity itself and not as a separate entry by default.

Same here. In fact I think we even agreed on this already at one of the previous meetings?!

Given ProtoBuf allows for the same sort of nesting as JSON/BSON does, the mapping really could and should like we do it for the document stores (MongoDB).

emmanuelbernard commented:
## Versioning and Optimistic locking

I remember you mentioned some specific feature that would benefit from this during our discussion, but I forgot. It was not about the exception vs double read for tuple creation. But something like that.

Sanne commented:
@gunnarmorling 

>> I think I would prefer to map associations within the entity itself and not as a separate entry by default.

> Same here. In fact I think we even agreed on this already at one of the previous meetings?!

Yes you all told me your preference, and I tried, and I'm sure I already explained several times that this wasn't going to work out, at least not without further patches in Infinispan.

To refresh your memory: the problem is not Protobuf, but ProtoStream. I switched 3 times between mapping strategies, also trying to bypass ProtoStream to accomodate for your mapping wishes, but ultimately the Infinispan team strongly suggested to stick to ProtoStream so I did fall back again to the Infinispan approach; The protostream library is meant to map user code directly and requires 1:1 mapping between encoders for a message and class model, so I can't nest things freely as one could by writing a Protobuf message with native APIs. On the other hand, Protobuf is meant to use code generation taking the schema as an input so we don't have many options.

N.B. each change in approach to follow "outsider suggestions" costed me weeks of work; the code now looks quite "clean" I hope.. or at least I tried hard for that, but this approach really is the only survivor of many attempts. I regularly shared the reasons of failed approaches so I'm not too happy to see you all shocked by the taken approach now.

All considered, I also think the strategy I adopted is safer for a general purpose default. I see the value in the other mapping strategy, but that needs to be controlled by an option so that implies you need to either have both options, or only the one I implemented. (The problem I'm thinking of is with huge collections: embedding it all wouldn't allow people to control loading with options such as ExtraLazy collection mappings ).

Having hopefully clear that the implemented mapping is needed too - at least as an option -, I see two possible evolutions to implement the second option (nesting of collections):
 - evolve ProtoStream (by either waiting for the Infinispan team or helping out)
 - bypass ProtoStream yet carefully mimic all of its storage idiosyncrasies (it is quite peculiar)

I suspect the best value is in evolving ProtoStream, yet that implies we limit our compatibility to some future release of Infinispan which has quite an impact on our roadmap plans.

So these are my consideration for having chosen the current mapping.

Sanne commented:
@emmanuelbernard RE 
## Bags & unordered collections

Good point I think your UUID idea could work; I opened the following issue to track this as a follow-up:

https://hibernate.atlassian.net/browse/OGM-1168

As suggested on the issue, I also think that unerdered bags should be stored embedded, at least by default. Maybe we should even take the option of alternative storage away in this case, as UUIDs can be costly and I suspect a bag would always need to be loaded as a whole anyway (hard to pick a subset of it).

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protostream/OgmProtoStreamMarshaller.java:
```diff
@@ -23,9 +23,7 @@
-	//FIXME all static threadlocals should die
-	// See https://hibernate.atlassian.net/browse/OGM-1163
-	private static final ThreadLocal<SerializationContext> currentSerializationContext = new ThreadLocal<>();
+	private final ThreadLocal<SerializationContext> currentSerializationContext = new ThreadLocal<>();
```
Why do you still need a thread local? `SerializationContext` is not concurrent safe?

Sanne replied:
There's only one instance of `OgmProtoStreamMarshaller` for the whole Hot Rod Client.

That means all our parallel Session are using the same marshaller - yet concurrently I need it to instruct which schema to use for a specific IO operation.

A typical Infinispan client app would have a different Class registered in the marshaller for each entity; we don't have this luxury, or I'd have to bytecode-generate a Class definition for each "Table" we want to map.

emmanuelbernard replied:
thanks, please add this as a comment in the code.

Sanne commented:
@emmanuelbernard RE
## Versioning and Optimistic locking

as mentioned in the PR description, some usage of Versioning was implemented but ultimately commented out as it needs to make sure the Association isn't deleted and then re-inserted when updated, as I lose track of version numbers at that point.

So this version does not implement optimistic locking.

emmanuelbernard commented on bom/pom.xml:
```diff
@@ -508,6 +521,52 @@
+                <scope>test</scope>
+            </dependency>
+            <dependency>
+                <groupId>org.infinispan</groupId>
```
why do we need the server? To run it during tests?

Sanne replied:
Yes: see RemoteHotRodServerRule

emmanuelbernard commented on core/src/main/java/org/hibernate/ogm/datastore/impl/DatastoreProviderType.java:
```diff
@@ -16,6 +16,7 @@
 public enum DatastoreProviderType {
 	MAP( "org.hibernate.ogm.datastore.map.impl.MapDatastoreProvider" ),
 	INFINISPAN( "org.hibernate.ogm.datastore.infinispan.impl.InfinispanDatastoreProvider" ),
```
Deprecate and add `INFINISPAN_EMBEDDED` ?

Sanne replied:
Ok: https://hibernate.atlassian.net/browse/OGM-1169

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java:
```diff
@@ -0,0 +1,396 @@
+/**
+ *  Some implementation notes for evolution:
+ *
+ *  - QueryableGridDialect can't be implemented as "native queries" in Hot Rod are DSL based
```
aren't the supported set of the Infinispan DSL anways a subset of JP-QL? We could ask people to use JP-QL anyways.

Sanne replied:
The IQL is not exposed yet, it will be in future versions (?)

emmanuelbernard replied:
yes but from the parser tree, you can call Infinispan DSL to build the query in the mean time. Granted it's not string -> string, it's string -> DSL object

Sanne replied:
Yes we could do that: not a technical issue. My concern is that we'd essentially be exposing IQL already before the datastore advertises (and stabilises) it?

Until Infinispan actually exposes it, you'd have OGM expose a capability which the store itself hasn't even yet documented.

emmanuelbernard replied:
Well, we do expose *JP-QL*. That's our job :)

Sanne replied:
Could you open a JIRA yourself in this case? I'm not sure I understood what you're proposing.

emmanuelbernard replied:
https://hibernate.atlassian.net/browse/OGM-1171

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java:
```diff
@@ -0,0 +1,396 @@
+ *
+ *  - OptimisticLockingAwareGridDialect can't be implemented as it requires an atomic replace
+ *    operation on a subset of the columns, while atomic operations in Hot Rod have to involve
+ *    either the full value or the version metadata of Infinispan's VersionedValue.
```
Can you explain what `VersionedValue` is for the uninitiated?

Sanne replied:
A read from Hot Rod:

    VersionedValue<ProtostreamPayload> v = client.getVersioned( key );
    ProtostreamPayload payload = v.getValue();
    long version = v.getVersion();

A write:

     boolean successfull = c.replaceWithVersion( key, payload, expectedVersion )

Alternatively, you can do a full comparison of the payload value:


     boolean success = c.replace( key, oldValue, newValue )

In this case I like the Infinispan API, but I don't think it matches the specific intent of the `OptimisticLockingAwareGridDialect` contract so optimistic locking would have to be implemented differently.

I'm not sure I understood the working of `OptimisticLockingAwareGridDialect` though. For example, it doesn't seem to have an "insertUnlessExists" semantics.

Sanne replied:
P.S. the critical difference is that the version is auto-incremented by Hot Rod, never explicitly written by the client.

DavideD replied:
I think the idea of this interface is to allow dialects without a native versioning system to still support optimistic locking using atomic find and update operations.

I suspect you can still implemente the insert in the uipdate method but I would need to check this.

Sanne replied:
@DavideD I agree. And since Infinispan has native support for it, I'll not implement the interface for the fallback option.

But also I'm having some concerns on the protocol proposed by `OptimisticLockingAwareGridDialect`.. I'm not sure that comparing all property values is correct as you might still fail to detect some forms of concurrent writes.

DavideD commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java:
```diff
@@ -0,0 +1,396 @@
+ *  - QueryableGridDialect can't be implemented as "native queries" in Hot Rod are DSL based
+ *    and have no String representation; this might change as Hot Rod exposes the underlying
+ *    query representation which is similar to HQL; alternatively we could look at exposing
+ *    native queries in some way other than a String.
```
I  don't think QueryableGridDialect expect a String representation, you can actually use a different object. For example, in MongoDB we have `MongoDBQueryDescriptor`.

I'm not saying that you have to implement it now, but I think the it can be done.

Sanne replied:
Which API is that .. I'm looking at

`org.hibernate.ogm.dialect.query.spi.QueryableGridDialect.parseNativeQuery(String)`

Also conceptually, the public exposed API is `org.hibernate.SharedSessionContract.createSQLQuery(String)`, right?

DavideD replied:
Yes, you are right, my bad.

I thought we had some way to pass something different since in `OgmSession` we have a `createNativeQuery` method but it is the same thing as `createSQLQuery`.

emmanuelbernard commented:
> To refresh your memory: the problem is not Protobuf, but ProtoStream. I switched 3 times between mapping strategies, also trying to bypass ProtoStream to accomodate for your mapping wishes, but ultimately the Infinispan team strongly suggested to stick to ProtoStream so I did fall back again to the Infinispan approach; The protostream library is meant to map user code directly and requires 1:1 mapping between encoders for a message and class model, so I can't nest things freely as one could by writing a Protobuf message with native APIs. On the other hand, Protobuf is meant to use code generation taking the schema as an input so we don't have many options.
> 
> N.B. each change in approach to follow "outsider suggestions" costed me weeks of work; the code now looks quite "clean" I hope.. or at least I tried hard for that, but this approach really is the only survivor of many attempts. I regularly shared the reasons of failed approaches so I'm not too happy to see you all shocked by the taken approach now.
> 
> All considered, I also think the strategy I adopted is safer for a general purpose default. I see the value in the other mapping strategy, but that needs to be controlled by an option so that implies you need to either have both options, or only the one I implemented. (The problem I'm thinking of is with huge collections: embedding it all wouldn't allow people to control loading with options such as ExtraLazy collection mappings ).

Don't get annoyed. Remember that we have not lived daily on this topic nor had weekly status update, peek. Assume in our discussions that we have forgotten *everything* (which is pretty much my case anyways).

We keep pushing for this because from a philosophy approach, we want OGM to store data like a user would have. And a user would have embedded the information.
So yes we can release this version as is but also we should push for a ProtoStream "dynamic" mode (my preferred approach over the alternatives you proposed).
How long it would take us to implement it ourselves for example (say asking a Yoann to go a run with it)?

DavideD commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java:
```diff
@@ -0,0 +1,396 @@
+	@Override
+	public void insertOrUpdateTuple(EntityKey key, Tuple tuple, TupleContext tupleContext) {
+		VersionedTuple versionedTuple = (VersionedTuple) tuple;
+		final long knownVersion = versionedTuple.getVersion();
```
The code that it is using this varibale is commented

Sanne replied:
thanks. It's needed by the block of code which is commented out below.. I'll remove all commented code and this.

Sanne commented:
I've re-pushed it to rebase and incorporate some of the feedback so far; especially improving comments and removed some dead code.

Had to disable the new tests GridDialectOperationInvocationForOneToOneTest and GridDialectOperationInvocationForElementCollectionTest as they fail after rebase.. I'll have to look at that.

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/InfinispanRemoteDatastoreProvider.java:
```diff
@@ -0,0 +1,192 @@
+		}
+	}
+
+	private void setMappedCacheNames(Set<String> tableNames) {
```
Never used.

Sanne replied:
thanks, removing.

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/InfinispanRemoteDatastoreProvider.java:
```diff
@@ -0,0 +1,192 @@
+
+	private SchemaOverride schemaOverrideService;
+
+	private Set<String> mappedCacheNames;
```
Not important but I think `mappedCacheNames` can be removed and one can use `perCacheSchemaMappers.keySet()`

Sanne replied:
`perCacheSchemaMappers.keySet()` is a subset of all mapped caches: we also use some for e.g. Sequences

Yet, I think you're right that I could change that: nothing would break as this is largely used just by testing utilities which don't care for Sequences.. but I like the explicitness of the Set in an explicit field, useful during debugging.

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteProperties.java:
```diff
@@ -0,0 +1,52 @@
+	 * The configuration property to use as key to define a custom configuration resource
+	 * for the Hot Rod (Infinispan remote) client.
+	 */
+	public static final String CONFIGURATION_RESOURCE_NAME = "hibernate.ogm.infinispan_remote.configuration_resource_name";
```
Does that means the generic host / port won't work and someone must provide some xml configuration file? Or is that optional? It might be good to guide users in the JavaDoc or maybe better doc as to what to prefer.

Sanne replied:
It's not an XML file but a properties file.
This file is mandatory to configure the Hot Rod client.

E.g. this is the one we use for testing:
 - https://github.com/Sanne/hibernate-ogm/blob/5f93b5bcf8974c412233ff260e817dfaa80d0062/infinispan-remote/src/test/resources/hotrod-client-testingconfiguration.properties

We could read such properties from the Hibernate properties? Last time I asked, you suggested to keep the external resource.

There's a follow up JIRA open to strictly validate the properties define in there, as we can't deal with some options being set.

emmanuelbernard replied:
@gunnarmorling for the other drivers like Neo or MongoDB don't we allow to pass through properties to the driver? We probably could align Infinispan's property injection as well by providing it a map from our own properties instead of using a file.

emmanuelbernard replied:
@Sanne ok, I think I disagree with younger @emmanuelbernard right now ;) At least we should do what we do for the other dialects.

gunnarmorling replied:
> for the other drivers like Neo or MongoDB don't we allow to pass through properties to the driver?

Yes, that's right. For MongoDB we all to pass through props like so:

    hibernate.ogm.mongodb.driver.serverSelectionTimeout=1000

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteProperties.java:
```diff
@@ -0,0 +1,52 @@
+	 * the configuration properties to override the Protobuf schema being generated.
+	 * This will not affect how entities are encoded, so the alternative schema must be compatible.
+	 */
+	public static final String SCHEMA_OVERRIDE_SERVICE = "hibernate.ogm.infinispan_remote.schema_override_service";
```
Since it's such an advanced feature, should we make it private?

Sanne replied:
I'm open to suggestions, but it seemed useful for end users.
They can use it to:

1. dump the generated schema to something else (a text file?)
2. intercept the generation and patch it before it's deployed
3. override the generation completely with their own resource

Of course no guarantees that the dialect will still work with a different schema than expected..

Sanne replied:
Obviously I created this for my own testing needs; tried to make it re-usable as secondary goal only.

gunnarmorling replied:
I'd say let's expose as few options as possible. At the very least mark it with `@Experimental`.

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protostream/ProtostreamPayload.java:
```diff
@@ -0,0 +1,98 @@
+		this.loadedSnapshot = null;
+	}
+
+	public void forEach(BiConsumer<String, Object> actionOnNamedTouples) {
```
s/touple/tuple/

gunnarmorling replied:
Or toupet ;)

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protostream/ProtostreamPayload.java:
```diff
@@ -0,0 +1,98 @@
+
+	public void forEach(BiConsumer<String, Object> actionOnNamedTouples) {
+		if ( loadedSnapshot != null ) {
+			loadedSnapshot.getMap().forEach( actionOnNamedTouples );
```
what does that do? Update the snapshot? Snapshots are supposed to be immutable I think.

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java:
```diff
@@ -0,0 +1,380 @@
+	public Tuple getTuple(EntityKey key, TupleContext tupleContext) {
+		final String cacheName = key.getTable();
+		ProtoStreamMappingAdapter mapper = provider.getDataMapperForCache( cacheName );
+		ProtostreamId idBuffer = mapper.createIdPayload( key.getColumnNames(), key.getColumnValues() );
```
Why is it called an idBuffer exactly?

emmanuelbernard commented:
For people who wonder while processing the PR, associations are stored in a cache by their table name. And each row is an full entry in the cache. So when loading an association, you essentially do a query. 

TODO @sanne do one to many clash with the owning entity table?
TODO That is very different from what we do for the other backends including Infinispan proper. Discuss with voice together to see if we are fine with that.

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java:
```diff
@@ -0,0 +1,380 @@
+					mapper.withinCacheEncodingContext( c -> c.remove( idBuffer ) );
+					break;
+				case CLEAR:
+					throw new AssertionFailure( "Request for CLEAR operation on an association mapped to dedicated entries. Makes no sense?" );
```
@sanne I think you got the wrong picture. It's not a clear of the association table. It's a clear of all the rowkey for that specific association entry (i.e. the assoc info dedicated to a specific entity).

So it would clear all entries returned by the query done in `getAssociation` (that's virtual of course and you can rely on the snapshot for that list).

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java:
```diff
@@ -0,0 +1,380 @@
+	public void insertOrUpdateAssociation(AssociationKey key, Association association, AssociationContext associationContext) {
+		final String cacheName = key.getTable();
+		final ProtoStreamMappingAdapter mapper = provider.getDataMapperForCache( cacheName );
+		final boolean embeddedInEntity = associatonTypeIsForeignKeyEmbeddedInEntity( key, association, mapper );
```
What does that mean exactly? The many to one case? or rather `AssociationType.ONE_TO_ONE`

Sanne replied:
Both of them, if they are the owners of the relation: as long as the foreign key to the linked entity is stored within this same table, and not in an external bridge table.

emmanuelbernard commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java:
```diff
@@ -0,0 +1,380 @@
+		}
+	}
+
+	private void insertOrUpdateAssociationEmbeddedInEntity(AssociationKey key, Association association, AssociationContext associationContext,
```
In this case if I understand correctly, you literally store all of the associaton columns (fk?) flat in the object.

```
User {
   String name;
   String addressId1;
   String addressId2;
}
// assuming Address is a composite key of id1 and id2
```

Correct?

In mongodb, we do create an intermediary object, something like

```
```
User {
   String name;
   AddressId address {
       String addressId1;
       String addressId2;
   }
}
// assuming Address is a composite key of id1 and id2
```

PS @gunnarmorling, I am correct on MongoDB right?

gunnarmorling replied:
Yes, for MongoDB a composite key would be [mapped into a sub-node](http://docs.jboss.org/hibernate/ogm/5.0/reference/en-US/html_single/#identifiers_3) in the document:

    {
      "_id" : {
          "author" : "Guillaume",
          "title" : "How to use Hibernate OGM ?"
      },
      "content" : "Simple, just like ORM but with a NoSQL database"
    }

Sanne replied:
Yes the storage is flat: remember I can't store nested structures.

emmanuelbernard replied:
Arrrrrrggggggggg, fuck ProtoStream and its flat structure limitation!

anistor replied:
@sanne I did not follow all the comments in this conversation, so why cannot you store nested structures? Protobuf/Protostream do allow it.

Sanne replied:
@anistor hey, I was going to ping you about this ;)
Remember our conversation in London? I'm constrained by the Marshaller API to require a 1:1 relation between class types and protobuf message names. We agreed we'd need to evolve that API, but in the meantime I decided to go ahead and implement this dialect w/o supporting nested structures yet. We'll talk again about this in Konstanz soon..

gunnarmorling commented:
> need to discuss to have a team understanding of the choices.

I suggest that we allocate an hour or so then in tomorrow morning's hangout for this.

gunnarmorling commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/BigDecimalProtofieldWriter.java:
```diff
@@ -0,0 +1,43 @@
+
+	public BigDecimalProtofieldWriter(int tag, String name, boolean nullable, String columnName) {
+		super( tag, name, nullable, columnName,
+				(ProtoStreamWriter outProtobuf, BigDecimal value) -> outProtobuf.writeBytes( name, value.toBigInteger().toByteArray() ),
```
The Lambdas look nice, but out of curiousity, is there any strong advantage over simply foreseeing methods in `ProtofieldWriter` to be implemented in the specific sub-classes? Just trying to figure when to use one over the other.

Sanne replied:
I think it's just verbosity, but there's also the aspect of the "name" variable being captured in the lambda bytecode. It effectively becomes a constant.

gunnarmorling commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java:
```diff
@@ -0,0 +1,380 @@
+	private final InfinispanRemoteDatastoreProvider provider;
+
+	public InfinispanRemoteDialect(InfinispanRemoteDatastoreProvider provider) {
+		this.provider = Objects.requireNonNull( provider );
```
No need to change from my side, but there is the `Contracts` utility which we've used for this in OGM so far.

Sanne replied:
This is the Java standard way; it wasn't available before Java 8

gunnarmorling commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/CompositeProtobufCoDec.java:
```diff
@@ -0,0 +1,185 @@
+import org.infinispan.protostream.MessageMarshaller.ProtoStreamReader;
+import org.infinispan.protostream.MessageMarshaller.ProtoStreamWriter;
+
+public final class CompositeProtobufCoDec implements MainOgmCoDec {
```
Why is it a "composite" codec? And why is it split into interface and impl?

Sanne commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/InfinispanRemoteDatastoreProvider.java:
```diff
@@ -0,0 +1,195 @@
+	private Set<String> mappedCacheNames;
+
+	//For each cache we have a schema and a set of encoders/decoders to the generated protobuf schema
+	private Map<String,ProtoDataMapper> perCacheSchemaMappers;
```
Nice, I wasn't aware. Will add this to several other locations too.

gunnarmorling commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protobuf/EnumProtofieldWriter.java:
```diff
@@ -0,0 +1,136 @@
+	private final int tag;
+	private final String name;
+	private final Class<? extends Enum> type;
+	private final Enum[] enumConstants;
```
Seems unused.

gunnarmorling commented on infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/impl/protostream/ProtoDataMapper.java:
```diff
@@ -0,0 +1,92 @@
+
+	public ProtoDataMapper(MainOgmCoDec delegate, SerializationContext serContext, OgmProtoStreamMarshaller marshaller) throws DescriptorParserException, IOException {
+		this.marshaller = marshaller;
+		this.delegate = Objects.requireNonNull( delegate );
```
Here we have access to the actual marshaller and the serialization context. Couldn't this be used to initialize the marshaller for once with the right context instead of setting this via the thread local again and again?

Sanne commented:
I've been applying the package renames and class renames as suggested, so my next diff will look terrible. Closing this PR..

Thanks everyone! Reopening soon on a fresh branch.

gunnarmorling commented on infinispan-remote/src/test/resources/protoschema-expectations/Cloud_SnowFlake.protobuf:
```diff
@@ -0,0 +1,40 @@
+}
+
+message Cloud {
+	required string id = 1;
```
We may have discussed it before, but did we decide on whether to include the id in the payload again or not? Is there a strong reason to do so?
